deeppavlov/datasets/robust_dataloader.py

```diff
@@ -0,0 +1,197 @@
+from collections import Counter
+
+
+class TextLoader:
+    def __init__(self, data_dir, batch_size, seq_length, chars=None, vocab=None):
+        self.data_dir = data_dir
+        self.batch_size = batch_size
+        self.seq_length = seq_length
+
+        self.chars = chars
+        self.vocab = vocab
+        self.noise_level = 0.0
+        vocab_file = os.path.join(data_dir, "vocab.pkl")
+        tensor_file = os.path.join(data_dir, "data.npy")
+        letter_vocab_file = os.path.join(data_dir, "letter_vocab.npy")
+
+        if not (os.path.exists(vocab_file)
+                and os.path.exists(tensor_file)
+                and os.path.exists(letter_vocab_file)):
+            print("reading text file")
+            self.preprocess(vocab_file, tensor_file, letter_vocab_file)
+        else:
+            print("loading preprocessed files")
+            self.load_preprocessed(vocab_file, tensor_file, letter_vocab_file)
+        self.create_batches()
+        self.reset_batch_pointer()
+
+    def preprocess(self, vocab_file, tensor_file, letter_vocab_file, ):
+        print("creating char vocab")
+        sents = self.create_vocab(vocab_file)
+
+        if self.vocab_size < 256:
+            dtype = np.uint8
+        else:
+            dtype = np.uint16
+
+        uniq_tokens = Counter()
+        word_sents = []
+        import re
+        for sent in sents:
+            sent = re.sub("\S*\d\S*", "", sent).strip()
+            sent.replace(".","")
+            sent.replace(":","")
+            word_s = word_tokenize(sent)
+            uniq_tokens.update(Counter(word_s))
+            word_sents.append(word_s)
+        count_pairs = sorted(uniq_tokens.items(), key=lambda x: -x[1])
+        tokens, _ = zip(*count_pairs)
+        tokens_vocab = dict(zip(tokens, range(len(tokens))))
+        letter_vectors = []
+        print("creating vocabs for w2v & letters")
+        for token in tqdm(tokens):
+            letter_vector = self._letters2vec(token, self.vocab, dtype)
+            letter_vectors.append(letter_vector)
+        self.letter_vocab = np.vstack(letter_vectors)
+        self.tensor = []
+        print("filling tensor")
+        for sent in tqdm(word_sents):
+            s = []
+            for t in sent:
+                s.append(tokens_vocab[t])
+            self.tensor.append(np.array(s, dtype=np.uint32))
+        self.word_vocab_size = len(uniq_tokens)
+        self.letter_size = self.letter_vocab.shape[1]
+        with codecs.open(tensor_file, "wb") as f:
+            cPickle.dump(self.tensor, f)
+        np.save(letter_vocab_file, self.letter_vocab)
+
+    def load_preprocessed(self, vocab_file, tensor_file, letter_vocab_file):
+        with codecs.open(vocab_file, 'rb') as f:
+            self.chars = cPickle.load(f)
+        self.vocab_size = len(self.chars) + 1
+        self.vocab = dict(zip(self.chars, range(1, len(self.chars) + 1)))
+        self.vocab[""] = 0
+        with codecs.open(tensor_file, "rb") as f:
```

Comment:
2. The function should be imported from deeppavlov.common.file

Reply:
which one?

Reply:
save_cpickle, load_cpickle - make these functions in deeppavlov.common.file and import from there